File: apps/demo-fe-elixir-phoenix/lib/demo_fe_exph/api/expenses.ex
defmodule DemoFeExph.Api.Expenses do
  @moduledoc "Expense API calls: CRUD operations and summary."

  alias DemoFeExph.Api.Client

  @doc "Lists expenses with optional pagination."
  def list_expenses(token, page \\ 1, size \\ 20) do
    query = URI.encode_query(%{page: page, size: size})
    Client.get("/api/v1/expenses?#{query}", token)
  end

  @doc "Fetches a single expense by ID."
  def get_expense(token, id) do
    Client.get("/api/v1/expenses/#{id}", token)
  end

  @doc "Creates a new expense with the given attributes."
  def create_expense(token, attrs) do
    Client.post("/api/v1/expenses", attrs, token)
  end

  @doc "Updates an existing expense by ID."
  def update_expense(token, id, attrs) do
    Client.put("/api/v1/expenses/#{id}", attrs, token)
  end

  @doc "Deletes an expense by ID."
  def delete_expense(token, id) do
    Client.delete("/api/v1/expenses/#{id}", token)
  end

  @doc "Returns the expense summary for the given currency."
  def get_expense_summary(token, currency \\ "USD") do
    query = URI.encode_query(%{currency: currency})
    Client.get("/api/v1/expenses/summary?#{query}", token)
  end
end

File: apps/demo-fe-ts-tanstackstart/app/lib/api/expenses.ts
import { apiFetch } from "./client";
import type { Expense, ExpenseListResponse, CreateExpenseRequest, UpdateExpenseRequest, ExpenseSummary } from "./types";

export function listExpenses(page = 0, size = 20): Promise<ExpenseListResponse> {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
  });
  return apiFetch<ExpenseListResponse>(`/api/v1/expenses?${params}`);
}

export function getExpense(id: string): Promise<Expense> {
  return apiFetch<Expense>(`/api/v1/expenses/${id}`);
}

export function createExpense(data: CreateExpenseRequest): Promise<Expense> {
  return apiFetch<Expense>("/api/v1/expenses", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function updateExpense(id: string, data: UpdateExpenseRequest): Promise<Expense> {
  return apiFetch<Expense>(`/api/v1/expenses/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export function deleteExpense(id: string): Promise<void> {
  return apiFetch(`/api/v1/expenses/${id}`, {
    method: "DELETE",
  });
}

export function getExpenseSummary(currency?: string): Promise<ExpenseSummary[]> {
  const params = new URLSearchParams();
  if (currency) params.set("currency", currency);
  return apiFetch<ExpenseSummary[]>(`/api/v1/expenses/summary?${params}`);
}
